internal/authentication/oidc/clientcredentials.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/oauth2"
+	ccreds "golang.org/x/oauth2/clientcredentials"
+
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	oidcScopeOpenID = "openid"
+)
+
+type ClientCredentialsOptions struct {
+	Logger       logger.Logger
+	TokenURL     string
+	ClientID     string
+	ClientSecret string
+	Scopes       []string
+	Audiences    []string
+	CAPEM        []byte
+}
+
+// ClientCredentials is an OAuth2 Token Source that uses the client_credentials
+// grant type to fetch a token.
+type ClientCredentials struct {
+	log          logger.Logger
+	currentToken *oauth2.Token
+	httpClient   *http.Client
+	fetchTokenFn func(context.Context) (*oauth2.Token, error)
+
+	lock sync.RWMutex
+}
+
+func NewClientCredentials(ctx context.Context, opts ClientCredentialsOptions) (*ClientCredentials, error) {
```

Comment:
Thinking about it, is this actually OIDC? Seems like we're just using OAuth2 here, as we are not concerned with the claims that are returned?

Reply:
Happy to add claim validation on the returned token if you think that makes sense?

Reply:
I don't think it's necessary TBH. The service will perform the validation when the token is presented, so us doing additional validation doesn't seem necessary here...

I am just thinking that maybe this package should be called "oauth2" rather than "oidc" since nothing here uses the OIDC standard, but just OAuth2.

Reply:
> I am just thinking that maybe this package should be called "oauth2" rather than "oidc" since nothing here uses the OIDC standard, but just OAuth2.

Why do you think this is the case?

Reply:
I don't see anything OIDC-specific in here. We are using the OAuth2 protocol to request an access token. We are not validating the claims contained in the token, so it doesn't need to be OIDC.

Pulsar wants token requested with the "openid" scope, but aside from that this should work as-is for all OAuth2 client credential flows?

Unless I'm missing something

Reply:
I think this is OIDC because we are fetching an identity token document, requested with the `openid` scope, which we are then using to _authenticate_ with a service on behalf of the "user". Remember that OIDC is built on top of OAuth2, and basically a constrained subset.

I think it is quite important to keep the distinction that we are using OIDC here- we are proving an identity to authenticate with the service, not given proof of authorization to access data. I'll add some validation to ensure the returned token has the right claims to be conformant.

Reply:
Alright, if we need OAuth2 we can always split it out in the future, creating a sub-class of this.